File: recursion/recursion-examples/java/palindrome.java
/*
Raspberry Pi Foundation
Developed as part of Ada Computer Science

Usage licensed under CC BY-NC-SA 4.0

Note: This file is designed to be copied out and compiled on your machine.
In order for the program to compile properly in some IDEs, you need to ensure that the
class filename is the same as the class name with the main method in it.

To run this file you need to:
1. Copy the contents
2. Create a Java project in the IDE of your choice (Eclipse, for example)
3. Create a new Java Class that uses the same name as the class containing the main method in this program
4. Paste the contents into the new Java Class
5. Save and run the program
*/

class RecursionExamples {
    // The main method is the entry point for all Java programs
    public static void main(String[] args) {
        String testWord = "kayak"; // Test data
        boolean isPalindrome = palindrome(testWord);
        System.out.println(testWord + ": " + isPalindrome);

        testWord = "apple"; // Test data
        isPalindrome = palindrome(testWord);
        System.out.println(testWord + ": " + isPalindrome);
    }


    // Returns True if string is a palindrome
    public static boolean palindrome(String word) {
        int length = word.length();
        if (length == 0 || length == 1) {
            return true;
        }
        else if (word.charAt(0) == word.charAt(length - 1)) {
            String newWord = word.substring(1, length - 1);
            return palindrome(newWord);
        }
        else {
            return false;
        }
    }


}

File: recursion/recursion-examples/python/palindrome.py
# Raspberry Pi Foundation
# Developed as part of Ada Computer Science
# 
# Usage licensed under CC BY-NC-SA 4.0

def palindrome(word):
    """Recursively determines whether a word is a palindrome. Returns a Boolean."""
    length = len(word)
    if length == 0 or length == 1:
        return True
    elif word[0] == word[length - 1]:
        new_word = word[1 : length - 1]
        return palindrome(new_word)
    else:
        return False


def main():
    test_word = "kayak"  # Test data
    is_palindrome = palindrome(test_word)
    print(f"{test_word}: {is_palindrome}")
    
    test_word = "apple"  # Test data
    is_palindrome = palindrome(test_word)
    print(f"{test_word}: {is_palindrome}")


# This code will run if this file is executed directly
# (i.e. not called by another program)
if __name__ == "__main__":
    main()
